src/core/exec-invoke.c

```diff
@@ -2548,6 +2557,157 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
         return 0;
 }
 
+static int set_exec_storage_quota(int fd, uint32_t proj_id, const QuotaLimit *ql) {
+        int r = 0;
+        uint64_t block_limit = 0;
+        uint64_t inode_limit = 0;
+
+        if (ql->quota_absolute == 0 || ql->quota_scale == 0) {
+                block_limit = 1;
+                inode_limit = 1;
+        } else if (ql->quota_absolute == UINT64_MAX) {
+                struct statvfs disk_st;
+                r = fstatvfs(fd, &disk_st);
+                if (r < 0)
+                        return r;
+
+                int percent = UINT32_SCALE_TO_PERCENT(ql->quota_scale);
+                block_limit = (uint64_t) DIV_ROUND_UP((disk_st.f_frsize * disk_st.f_blocks / 100) * percent, QIF_DQBLKSIZE);
+                inode_limit = (uint64_t) ((disk_st.f_files / 100) * percent);
```

Comment:
instead of covnerting to percent first, let's just operate directly with the greater accuracy that uint32_t normalization gives you?